src/Microsoft.Health.Fhir.Api/Features/Throttling/ThrottlingMiddleware.cs

```diff
@@ -0,0 +1,69 @@
+﻿// -------------------------------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using EnsureThat;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Fhir.Api.Configs;
+using Microsoft.Health.Fhir.Api.Features.ActionResults;
+
+namespace Microsoft.Health.Fhir.Api.Features.Throttling
+{
+    public class ThrottlingMiddleware
+    {
+        private readonly ILogger<ThrottlingMiddleware> _logger;
+        private readonly ThrottlingConfiguration _configuration;
+        private int _requestsInFlight = 0;
+        private RequestDelegate _next;
+
+        public ThrottlingMiddleware(RequestDelegate next, IOptions<ThrottlingConfiguration> configuration, ILogger<ThrottlingMiddleware> logger)
+        {
+            _next = EnsureArg.IsNotNull(next, nameof(next));
+            _logger = EnsureArg.IsNotNull(logger, nameof(logger));
+            _configuration = EnsureArg.IsNotNull(configuration?.Value, nameof(configuration));
+        }
+
+        public async Task Invoke(HttpContext context)
+        {
+            if (_configuration.ExcludedEndpoints.Contains($"{context.Request.Method}:{context.Request.Path.Value?.TrimEnd('/')}", StringComparer.OrdinalIgnoreCase))
+            {
+                // Endpoint is exempt from concurrent request limits.
+                await _next(context);
+                return;
+            }
+
+            try
+            {
+                if (Interlocked.Increment(ref _requestsInFlight) <= _configuration.ConcurrentRequestLimit)
+                {
+                    // Still within the concurrent request limit, let the request through.
+                    await _next(context);
+                }
+                else
+                {
+                    // Exceeded the concurrent request limit, return 429.
+                    context.Response.StatusCode = (int)HttpStatusCode.TooManyRequests;
+                    _logger.LogWarning($"{Resources.TooManyConcurrentRequests}. Limit is {_configuration.ConcurrentRequestLimit}.");
+
+                    // Output an OperationOutcome in the body.
+                    var result = TooManyRequestsActionResult.TooManyRequests;
+                    await result.ExecuteResultAsync(new ActionContext { HttpContext = context, RouteData = context.GetRouteData() ?? new RouteData() });
```

Comment:
I feel like this is sort of mixing in MVC stuff before MVC has even been reached, but without the full behavior we would get if this were executed later in the pipeline. For instance, we return this body:
```
{
    "severity": "Error",
    "code": "Throttled",
    "diagnostics": "The maximum number of concurrent API calls was reached.",
    "location": null
}
```

(And we ignore if the request requested XML)

Whereas we return this for a 404:
```
{
    "resourceType": "OperationOutcome",
    "id": "d4803fe1-6bd9-44ea-a635-310ddb3ea92a",
    "issue": [
        {
            "severity": "error",
            "code": "not-found",
            "diagnostics": "Resource type 'Patient' with id 'sss' couldn't be found."
        }
    ]
}
```

I think it is completely fine in this case, maybe I'm making a general comment about our codebase design and the reliance on MVC.
Also, given this, I wonder if we should just write a cached  binary serialization to the response stream. I don't know what the ovehead of the `ExecuteResultAsync` call with its argument payload is.